File: backend/app/strategy/mean_reversion.py
"""Bollinger Band mean-reversion: fade a close outside the bands back to the mean.

Best suited to ranging regimes -- the scoring engine's trend_alignment factor
naturally penalizes this strategy's signals when the regime is strongly
trending, since a reversion trade against a strong trend is exactly the
"fights the trend" case.
"""
from __future__ import annotations

from decimal import Decimal

import pandas as pd

from app.strategy.base import Signal, Strategy

PERIOD = 20
NUM_STD = 2.0


class MeanReversionStrategy(Strategy):
    strategy_id = "mean_reversion_bollinger_20"

    def generate_signal(self, symbol: str, df: pd.DataFrame) -> Signal | None:
        if len(df) < PERIOD + 1:
            return None

        closes = df["close"]
        sma = closes.rolling(PERIOD).mean().iloc[-1]
        std = closes.rolling(PERIOD).std(ddof=0).iloc[-1]
        if pd.isna(sma) or pd.isna(std) or std == 0:
            return None

        upper = float(sma + NUM_STD * std)
        lower = float(sma - NUM_STD * std)
        last = df.iloc[-1]
        close = float(last["close"])

        if close < lower:
            return Signal(
                strategy_id=self.strategy_id,
                symbol=symbol,
                side="long",
                structure_swing_price=Decimal(str(float(last["low"]))),
                reason=f"close {close} is below the lower Bollinger Band ({lower:.2f}), reversion toward {sma:.2f} expected",
            )
        if close > upper:
            return Signal(
                strategy_id=self.strategy_id,
                symbol=symbol,
                side="short",
                structure_swing_price=Decimal(str(float(last["high"]))),
                reason=f"close {close} is above the upper Bollinger Band ({upper:.2f}), reversion toward {sma:.2f} expected",
            )
        return None

File: backend/src/strategy/meanReversion.ts
/**
 * Bollinger Band mean-reversion: fade a close outside the bands back to the mean.
 *
 * Best suited to ranging regimes -- the scoring engine's trendAlignment
 * factor naturally penalizes this strategy's signals when the regime is
 * strongly trending, since a reversion trade against a strong trend is
 * exactly the "fights the trend" case.
 */
import { Decimal } from "decimal.js";
import type { OhlcBar } from "../regime/indicators.js";
import type { Signal, Strategy } from "./types.js";

const PERIOD = 20;
const NUM_STD = 2.0;

function mean(xs: number[]): number {
  return xs.reduce((a, b) => a + b, 0) / xs.length;
}

export class MeanReversionStrategy implements Strategy {
  strategyId = "mean_reversion_bollinger_20";

  generateSignal(symbol: string, bars: OhlcBar[]): Signal | null {
    if (bars.length < PERIOD + 1) return null;

    const closes = bars.slice(-PERIOD).map((b) => b.close);
    const sma = mean(closes);
    const std = Math.sqrt(mean(closes.map((c) => (c - sma) ** 2)));
    if (!std) return null;

    const upper = sma + NUM_STD * std;
    const lower = sma - NUM_STD * std;
    const last = bars[bars.length - 1]!;

    if (last.close < lower) {
      return {
        strategyId: this.strategyId,
        symbol,
        side: "long",
        structureSwingPrice: new Decimal(last.low),
        reason: `close ${last.close} is below the lower Bollinger Band (${lower.toFixed(2)}), reversion toward ${sma.toFixed(2)} expected`,
      };
    }
    if (last.close > upper) {
      return {
        strategyId: this.strategyId,
        symbol,
        side: "short",
        structureSwingPrice: new Decimal(last.high),
        reason: `close ${last.close} is above the upper Bollinger Band (${upper.toFixed(2)}), reversion toward ${sma.toFixed(2)} expected`,
      };
    }
    return null;
  }
}
